cmd/ci-secret-generator/main.go

```diff
@@ -0,0 +1,226 @@
+	bytes, err := ioutil.ReadFile(o.bwPasswordPath)
+	if err != nil {
+		return err
+	}
+	o.bwPassword = strings.TrimSpace(string(bytes))
+	secrets.Insert(o.bwPassword)
+
+	bytes, err = ioutil.ReadFile(o.configPath)
+	if err != nil {
+		return err
+	}
+
+	err = yaml.Unmarshal(bytes, &o.config)
+	if err != nil {
+		return err
+	}
+	return o.validateCompletedOptions()
+}
+
+func (o *options) validateCompletedOptions() error {
+	if o.bwPassword == "" {
+		return fmt.Errorf("--bw-password-file was empty")
+	}
+
+	for i, bwItem := range o.config {
+		if bwItem.ItemName == "" {
+			return fmt.Errorf("config[%d].itemName: empty key is not allowed", i)
+		}
+		switch bwItem.Attribute.Name {
+		case attributeTypePassword, "":
+		default:
+			return fmt.Errorf("config[%d].attribute: only the '%s' is supported, not %s", i, attributeTypePassword, bwItem.Attribute.Name)
+		}
+		if (bwItem.Field.Name != "" && bwItem.Field.Cmd == "") ||
+			(bwItem.Attribute.Name != "" && bwItem.Attribute.Cmd == "") ||
+			(bwItem.Attachment.Name != "" && bwItem.Attachment.Cmd == "") {
+			return fmt.Errorf("config[%d]: empty field not allowed for cmd if name is specified for any of attribute, field or attachment", i)
+		}
+	}
+	return nil
+}
+
+func executeCommand(command string) ([]byte, error) {
+	cmd := strings.Fields(command)
+	out, err := exec.Command(cmd[0], cmd[1:]...).CombinedOutput()
+	if err != nil {
+		return nil, fmt.Errorf("bw cmd failed: %v", string(out))
+	}
+	return out, nil
+}
+
+func processItem(w io.Writer, cmd string, entryType string, dryRun bool, procesor func([]byte) error) error {
+	out, err := executeCommand(cmd)
+	if err != nil {
+		return fmt.Errorf("failed to execute command: %w", err)
+	}
+	if dryRun {
+		if _, err := fmt.Fprintf(w, "\t%s: %s", entryType, string(out)); err != nil {
+			return err
+		}
+	} else {
+		return procesor(out)
+	}
+	return nil
+}
+
+func updateSecrets(bwItems []bitWardenItem, bwClient bitwarden.Client, dryRun bool) error {
+	if dryRun {
+		logrus.Infof("Dry-Run enabled, writing secrets to stdout")
+		if _, err := fmt.Fprintln(os.Stdout, "---"); err != nil {
+			return err
+		}
+	}
+	for _, bwItem := range bwItems {
+		if dryRun {
+			if _, err := fmt.Fprintf(os.Stdout, "Item: %s", bwItem.ItemName); err != nil {
+				return err
+			}
+		}
+		if bwItem.Field.Name != "" {
+			if err := processItem(os.Stdout, bwItem.Field.Cmd, "Field", dryRun, func(out []byte) error {
+				if err := bwClient.SetFieldOnItem(bwItem.ItemName, bwItem.Field.Name, out); err != nil {
+					return fmt.Errorf("failed to set field item: %s, field: %s - %w", bwItem.ItemName, bwItem.Field.Name, err)
+				}
+				return nil
+			}); err != nil {
+				return err
+			}
+		}
+		if bwItem.Attachment.Name != "" {
+			if err := processItem(os.Stdout, bwItem.Attachment.Cmd, "Attachment", dryRun, func(out []byte) error {
+				if err := bwClient.SetAttachmentOnItem(bwItem.ItemName, bwItem.Attachment.Name, out); err != nil {
```

Comment:
Rather than passing in a func that wraps the `bwClient`, why not directly pass in the `bwClient` as an interface and use a mock-implementation for dry run and tests?

Reply:
bwClient is already an interface, and I could probably change the Run function to make it print the commands to be executed to a file for dry run. But that would be the 64bit encoded data (the output of the command wouldn't be human readable then but would be exactly the command executed )
Or as you mentioned write a separate mock implementation to just spit out the data into a file.
Which would be better?

Reply:
A seperate mock implementation